ibc-testkit/src/hosts/mod.rs

```diff
@@ -1 +1,85 @@
-pub mod block;
+use core::fmt::Debug;
+
+use ibc::core::client::types::Height;
+use ibc::core::host::types::identifiers::ChainId;
+use ibc::core::primitives::prelude::*;
+use ibc::core::primitives::Timestamp;
+use ibc::primitives::proto::Any;
+
+use crate::testapp::ibc::clients::{AnyClientState, AnyConsensusState};
+
+pub mod mock;
+pub mod tendermint;
+
+pub use mock::Host as MockHost;
+pub use tendermint::Host as TendermintHost;
+
+/// TestHost is a trait that defines the interface for a host blockchain.
+pub trait TestHost: Debug {
+    /// The type of block produced by the host.
+    type Block: TestBlock;
+
+    /// The type of client state produced by the host.
+    type ClientState: Into<AnyClientState>;
+
+    /// The type of block parameters to produce a block
+    type BlockParams: Debug + Default;
+
+    /// The type of light client parameters to produce a light client state
+    type LightClientParams: Debug + Default;
+
+    /// Create a new host with the given chain identifier.
+    fn with_chain_id(chain_id: ChainId) -> Self;
+
+    /// The chain identifier of the host.
+    fn chain_id(&self) -> &ChainId;
+
+    /// Generate a block at the given height and timestamp, using the provided parameters.
+    fn generate_block(
+        &self,
+        height: u64,
+        timestamp: Timestamp,
+        params: &Self::BlockParams,
+    ) -> Self::Block;
+
+    /// Generate a client state using the block at the given height and the provided parameters.
+    fn generate_client_state(
+        &self,
+        latest_block: &Self::Block,
+        params: &Self::LightClientParams,
+    ) -> Self::ClientState;
+}
+
+/// TestBlock is a trait that defines the interface for a block produced by a host blockchain.
+pub trait TestBlock: Clone + Debug {
+    /// The type of header can be extracted from the block.
+    type Header: TestHeader + From<Self>;
+
+    /// The height of the block.
+    fn height(&self) -> Height;
+
+    /// The timestamp of the block.
+    fn timestamp(&self) -> Timestamp;
```

Comment:
Could default implementations for these two methods be provided using the `Header`?
It even might be worthwhile to explore whether merging `TestBlock` and `TestHeader` under a single trait. Uncertain if such granularity is necessary.

Reply:
I wanted to differentiate between `TestBlock` and `TestHeader` in the Tendermint context. The Tendermint light client header contains _trusted_ height metadata. Let me think more about this.